src/core/components/pin-set.js

```diff
@@ -0,0 +1,267 @@
+        cb = once(cb)
+        const pbHeader = pb.Set.encode({
+          version: 1,
+          fanout: defaultFanout,
+          seed: depth
+        })
+        const headerBuf = Buffer.concat([
+          Buffer.from(varint.encode(pbHeader.length)), pbHeader
+        ])
+        const fanoutLinks = []
+        for (let i = 0; i < defaultFanout; i++) {
+          fanoutLinks.push(new DAGLink('', 1, emptyKey))
+        }
+
+        if (pins.length <= maxItems) {
+          const nodes = pins
+            .map(item => ({
+              link: new DAGLink('', 1, item.key),
+              data: item.data || Buffer.alloc(0)
+            }))
+            // sorting makes any ordering of `pins` produce the same DAGNode
+            .sort((a, b) => Buffer.compare(a.link.multihash, b.link.multihash))
+
+          const rootLinks = fanoutLinks.concat(nodes.map(item => item.link))
+          const rootData = Buffer.concat(
+            [headerBuf].concat(nodes.map(item => item.data))
+          )
+
+          DAGNode.create(rootData, rootLinks, (err, rootNode) => {
+            if (err) { return cb(err) }
+            return cb(null, rootNode)
+          })
+        } else {
+          // If the array of pins is > maxItems, we:
+          //  - distribute the pins among `defaultFanout` bins
+          //    - create a DAGNode for each bin
+          //      - add each pin of that bin as a DAGLink
+          //  - create a root DAGNode
+          //    - store each bin as a DAGLink
+          //  - send that root DAGNode via `cb`
+          // (using go-ipfs' "wasteful but simple" approach for consistency)
+
+          const bins = pins.reduce((bins, pin) => {
+            const n = hash(depth, pin.key) % defaultFanout
+            bins[n] = n in bins ? bins[n].concat([pin]) : [pin]
+            return bins
+          }, {})
+
+          const binKeys = Object.keys(bins)
+          binsToFill += binKeys.length
+          binKeys.forEach(n => {
+            storePins(
+              bins[n],
+              (err, child) => storeChild(err, child, n),
+              depth + 1,
+              binsToFill,
+              binsFilled
+            )
+          })
+        }
+
+        function storeChild (err, child, bin) {
+          if (err) { return cb(err) }
+
+          const cid = new CID(child._multihash)
+          dag.put(child, { cid }, (err) => {
+            if (err) { return cb(err) }
+
+            fanoutLinks[bin] = new DAGLink('', child.size, child.multihash)
+            binsFilled++
+
+            if (binsFilled === binsToFill) {
+              // all finished
+              DAGNode.create(headerBuf, fanoutLinks, (err, rootNode) => {
+                if (err) { return cb(err) }
+                return cb(null, rootNode)
+              })
+            }
+          })
+        }
+      }
+    },
+
+    loadSet: (rootNode, name, callback) => {
+      callback = once(callback)
```

Comment:
I don't think we need a `once` here unless it's possible for `pinSet.walkItems` to call the callback twice (which it shouldn't do)